### PR TITLE
Exporting LLVM CPU functions when debug symbols are enabled.

### PR DESCRIPTION
**Pull request by benvanik**
The exports are not used and we normally omit them to reduce binary size. When disassembling/decompiling, though, having the exported symbols makes it much easier to find the dispatches and correlate them to the original program.

Example:
image

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp [hunk 1]
```diff
@@ -410,13 +412,38 @@ class LLVMCPUTargetBackend final : public TargetBackend {
...
       global.setDSOLocal(true);
       global.setLinkage(llvm::GlobalValue::LinkageTypes::InternalLinkage);
     }
 
+    // Since all exports are fetched via the executable library query function
+    // we don't need to make them publicly exported on the module. This has
+    // the downside of making some tooling (disassemblers/decompilers/binary
+    // size analysis tools/etc) a bit harder to work with, though, so when
+    // compiling in debug mode we export all the functions.
+    if (options_.debugSymbols) {
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:430]: I wonder if we should be turning this off for benchmarks or call it out somewhere in user docs...

https://github.com/iree-org/iree/blob/56a873934c70a440e26f39ae14eb82ba0ed54897/compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h#L43-L47 https://github.com/iree-org/iree/blob/56a873934c70a440e26f39ae14eb82ba0ed54897/compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp#L155-L159

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:430]: Should we store debug symbols off by default, and enable them in IREEs debug builds?

Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:430]: We'd want the opposite - this should be _on_ for all benchmarks - nothing here impacts runtime performance (or shouldn't), just provides more information for the kind of tools that'd be used in benchmarking.

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp [hunk 2]
```diff
@@ -410,13 +412,38 @@ class LLVMCPUTargetBackend final : public TargetBackend {
         continue;
       }
       func.setDSOLocal(true);
-      func.setLinkage(llvm::GlobalValue::LinkageTypes::InternalLinkage);
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:413]: Some CI errors for this

https://github.com/iree-org/iree/actions/runs/3989611477/jobs/6842259299
```
[702/821] Generating check_llvm-cpu_local-task_floor.mlir_module.vmfb from floor.mlir
FAILED: tests/e2e/xla_ops/check_llvm-cpu_local-task_floor.mlir_module.vmfb /work/full-build-dir/tests/e2e/xla_ops/check_llvm-cpu_local-task_floor.mlir_module.vmfb 
cd /work/full-build-dir/tests/e2e/xla_ops && /work/full-build-dir/tools/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=mhlo /work/tests/e2e/xla_ops/floor.mlir -o check_llvm-cpu_local-task_floor.mlir_module.vmfb --iree-hal-executable-object-search-path=\"/work/full-build-dir\" --iree-llvm-embedded-linker-path=\"/work/full-build-dir/third_party/llvm-project/llvm/bin/lld\" --iree-llvm-wasm-linker-path=\"/work/full-build-dir/third_party/llvm-project/llvm/bin/lld\"
lld: error: undefined symbol: floorf
```

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp (hunk 1 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:430]: Ah, okay. Default to enabling debug symbols, only disable debug symbols when trying to minimize binary size? Does the "significantly change the output program and benchmarking" comment only apply when setting debug symbols separately from opt level?

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp (hunk 2 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp:413]: bah thanks for pointing this out! this stuff is so fiddly x_x

**Comment by benvanik** [comment]
There's one failure of `iree/hal/drivers/local_sync/cts/local-sync_embedded-elf_command_buffer_dispatch_test`+local-task only in TSAN which confusingly is using the _system_ loader instead of the _embedded-elf_ loader as TSAN requires system linking. I'm going to pull back what this PR is doing to just make Windows and embedded ELFs expose the symbols properly for now and leave linux/etc for someone else to hopefully figure out.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**